### PR TITLE
Bind Window aggregate expressions to system.main

### PR DESCRIPTION
**Pull request by Tmonster**
Currently if the window catalog and schema are left empty, it is not guaranteed that the catalog function used for the window is an aggregate function. If the user creates a macro with the same name as an aggregate function (as suggested in https://github.com/duckdb/duckdb/issues/5832), then the macro can be returned. Since macros can only define scalar functions*, this results in an error. The  fix for this is if a window expression doesn't have a catalog or schema, set them to the system catalog and the default schema. Since that is the only place where aggregate functions can be created, this problem should no longer be an issue

I can imagine there might be better places to put this check. 

*Assumed because of the existence of scalar_macro_function.hpp but no aggregate_macro_function.hpp
Also the fact that MacroFunction has a `MacroType type` enum attribute which only has the following values 
```
enum class MacroType : uint8_t { VOID_MACRO = 0, TABLE_MACRO = 1, SCALAR_MACRO = 2 };
```

### CODE REVIEW

File: src/planner/binder/expression/bind_window_expression.cpp [hunk 1]
```diff
@@ -192,6 +192,10 @@ BindResult SelectBinder::BindWindow(WindowExpression &window, idx_t depth) {
 	unique_ptr<FunctionData> bind_info;
 	if (window.type == ExpressionType::WINDOW_AGGREGATE) {
 		//  Look up the aggregate function in the catalog
+		if (window.catalog.empty() && window.schema.empty()) {
+			window.catalog = SYSTEM_CATALOG;
+			window.schema = DEFAULT_SCHEMA;
+		}
 		auto func = Catalog::GetEntry<AggregateFunctionCatalogEntry>(context, window.catalog, window.schema,
 		                                                             window.function_name, false, error_context);
 		D_ASSERT(func->type == CatalogType::AGGREGATE_FUNCTION_ENTRY);
```
Comment (Mytherin) [inline src/planner/binder/expression/bind_window_expression.cpp:201]: Perhaps this should be an exception instead of an assertion?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! As mentioned this might cause problems as this would essentially negate the macro for window functions (so that Window functions would use the standard sum behavior, returning NULL instead of 0 for empty sets). Perhaps we should move away from the macro approach for this functionality after all?

File: src/planner/binder/expression/bind_window_expression.cpp (hunk 1 above)
Comment (Tmonster) [reply src/planner/binder/expression/bind_window_expression.cpp:201]: In Catalog::GetEntry an error is thrown if the function is not of type aggregate 
https://github.com/duckdb/duckdb/blob/master/src/include/duckdb/catalog/catalog.hpp#L213

**Comment by Mytherin** [comment]
In that case I would say the current behavior is fully as expected - and we should fix this in another way (by e.g. creating a different `sum` implementation as an aggregate function in the R client).

**Comment by Tmonster** [comment]
That's a good point. This is an edge case the will require more work I think. Would be nice if users could define certain characteristics for aggregate functions via a MACRO. Maybe an initial value of 0 or NULL.

**Comment by Tmonster** [comment]
> In that case I would say the current behavior is fully as expected - and we should fix this in another way (by e.g. creating a different `sum` implementation as an aggregate function in the R client).

Yea, I don't think this PR has the correct solution. If a column has only NULL entries, the output should be 0 (in accordance to the macro), but the system.main sum function will return NULL. 

So to rephrase my idea from above, if users could choose the starting value for the aggregate sum function (either 0, or NULL), then this should solve this problem. Currently the catalog sum function treats NULL as 0 in scalar operations, so 0 + NULL = 0, but NULL + NULL = NULL.

Maybe this could be a pragma `sum_initial_value` or so?

**Comment by Mytherin** [comment]
Perhaps the R client can override the `aggregate_finalize_t finalize` function pointer of the `sum` function in the catalog?

Untested but something like this:

```cpp
struct ZeroSumOperation {
	template <class T, class STATE>
	static void Finalize(Vector &result, AggregateInputData &, STATE *state, T *target, ValidityMask &mask, idx_t idx) {
		target[idx] = Hugeint::Convert(state->value);
	}
};

...
// init
auto sum_function = catalog.GetCatalogEntry<AggregateCatalogEntry>(context, SYSTEM_CATALOG, DEFAULT_SCHEMA, "sum");

for(auto &aggr : sum_function->functions.functions) {
   	switch(aggr.arguments[0].InternalType()) {
   	case PhysicalType::INT64:
	    aggr.finalize = AggregateFunction::StateFinalize<SumState<int64_t>, int64_t, ZeroSumOperation>;
	    break;
   	}
}
...

```

**Comment by Mytherin** [comment]
Closing this pending a different solution

**Pull request closed without merging**